cmd/backup/main.go

```diff
@@ -65,4 +82,83 @@ func main() {
 	s.must(s.withLabeledCommands(lifecyclePhaseProcess, s.encryptArchive)())
 	s.must(s.withLabeledCommands(lifecyclePhaseCopy, s.copyArchive)())
 	s.must(s.withLabeledCommands(lifecyclePhasePrune, s.pruneBackups)())
+
+	return nil
+}
+
+func main() {
+	serve := flag.Bool("foreground", false, "run the backup in the foreground")
+	envFolder := flag.String("env-folder", "/etc/dockervolumebackup/conf.d", "location of environment files")
+	flag.Parse()
+
+	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
+
+	if *serve {
+		logger.Info("running in the foreground instead of cron")
+
+		cr := cron.New()
+
+		addJob := func(c *Config) {
+			logger.Info("added cron job", "schedule", c.BackupCronExpression)
+			_, err := cr.AddFunc(c.BackupCronExpression, func() {
+				err := runBackup(c)
+				if err != nil {
+					logger.Error("unexpected error during backup", "error", err)
+				}
+			})
+			if err != nil {
+				logger.Error("failed to create cron job", "schedule", c.BackupCronExpression)
+			}
+		}
+
+		cs, err := loadEnvFiles(*envFolder)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				logger.Error("could not load config from environment files")
+				os.Exit(1)
+			}
+
+			c, err := loadEnvVars()
+			if err != nil {
+				logger.Error("could not load config from environment variables")
+				os.Exit(1)
+			} else {
+				addJob(c)
+			}
+		} else {
+			for _, c := range cs {
+				addJob(c)
+			}
+		}
+
+		logger.Info("subscribed to interupt signals")
+		var quit = make(chan os.Signal, 1)
+		signal.Notify(quit, syscall.SIGTERM, syscall.SIGINT)
+
+		logger.Info("starting cron scheduler")
+		cr.Start()
+
+		logger.Info("application goes to sleep")
+		<-quit
+
+		logger.Info("interrupt arrived, stopping schedules")
+		ctx := cr.Stop()
+		<-ctx.Done()
+	} else {
+		logger.Info("executing one time backup")
+
+		c, err := loadEnvVars()
+		if err != nil {
+			logger.Info("could not load config from environment variables", "error", err)
+			os.Exit(1)
```

Comment:
I think we can improve that `func main` now has to manually exit in so many places. Previously, we had the deferred `recover` that would see all fatal errors and would be the only place where the code force exits the program.

I think we should introduce something similar here as well, i.e. another `must` function that is responsible for handling errors that happen in `func main`.

Maybe, it'd even be a good idea to start adding an `app` type that kind of works like `script`, but is responsible for managing collecting all dependencies of the application when running in the foreground.

Reply:
For now I only created a wrapper `app` function that can return an `error`. I think it will be enough, there should be no panic between the recover handler in script and the main function.

Reply:
I'll need some more focused time to look at this, so I'll have a look when I can and will feedback on this particular topic.